external/js/cothority/lib/identity.js

```diff
@@ -0,0 +1,227 @@
+"use strict";
+
+const topl = require('topl');
+const UUID = require('pure-uuid');
+const protobuf = require('protobufjs')
+
+const misc = require('./misc');
+
+
+/**
+ * ServerIdentity represents a cothority server. It has an associated public key
+ * and a TCP address. The WebSocket address is derived from the TCP address.
+ * */
+class ServerIdentity {
+    /*
+     * Returns a new ServerIdentity from the public key and address.
+     * @param {Uint8Array} public key in bytes format
+     * @param {string} address tcp address of the cothority node
+     * @param {string} description of the conode. Can be null.
+     * @return a ServerIdentity
+     * */
+    constructor(publicKey, address,description) {
+        if (publicKey.constructor !== Uint8Array)
+            throw TypeError;
+        this.pub = publicKey
+        this.addr = address;
+        this._description = description;
+        // id of the identity
+        const url = 'https://dedis.epfl.ch/id/' + misc.uint8ArrayToHex(this.pub);
+        this._id = new UUID(5, 'ns:URL', url).export();
+        // tcp + websocket address
+        let parts = address.replace("tcp://", "").split(":");
+        parts[1] = parseInt(parts[1]) + 1;
+        this.wsAddr = "ws://" + parts.join(":");
+    }
+
+    /*
+     * Returns a new ServerIdentity from the public key in hexadecimal format
+     * and address
+     * @param {string} hex-encoded public key
+     * @param {string} address
+     * @return a ServerIdentity
+     * */
+    static fromHexPublic(hexPublic,address,description) {
+        const pub = misc.hexToUint8Array(hexPublic);
+        return new ServerIdentity(pub,address,description);
+    }
+    /*
+     * @return the public key as a Uint8Array buffer
+     * */
+    get public() {
+        return this.pub;
+    }
+
+    /*
+     * @return the WebSocket address. That can be passed into the net.Socket
+     * address constructor argument.
+     * */
+    get websocketAddr() {
+        return this.wsAddr;
+    }
+
+    /*
+     * @return the tcp address of the server
+     * */
+    get tcpAddr() {
+        return this.addr;
+    }
+
+    /*
+     * @return the id of this serveridentity
+     * */
+    get id() {
+        return this._id;
+    }
+
+    /**
+     * returns the description associated with this identity
+     *
+     * @returns {string} the string description
+     */
+    get description() {
+        return this._description;
+    }
+
+    toString() {
+        return this.tcpAddr;
+    }
+
+    /**
+     * point returns the point representation of the public key
+     *
+     * @param {kyber.Group} group the expected group from which the public key belongs
+     * @returns {kyber.Point} the kyber.Point representing the public key
+     */
+    point(group) {
```

Comment:
It seems strange to postpone identifying the group until here. Why doesn't the constructor take a point as it's first argument?